### PR TITLE
Unbias generation of requests.

### LINKED ISSUE #18: Bias towards the first endpoints
The `HasGenRequest` instance for `:<|>` is nice and simple but leads to an unsatisfactory bias.

This can be demonstrated with the following snippet:

```
λ> map length . group . sort <$> sequence (replicate 100 (generate (oneof [pure 1, pure 2, pure 3])))
[33,30,37]
λ> map length . group . sort <$> sequence (replicate 100 (generate (oneof [pure 1, oneof [pure 2, pure 3]])))
[54,20,26]
```

It would be nice if alternatives where _effectively_ associatives.

### PR DESCRIPTION
**Pull request by jkarni**
So that each endpoint is picked with the same frequency.
Fixes #18

### CODE REVIEW

**Comment by jkarni** [comment]
Still WIP

**Comment by erewok** [comment]
Need to add a test for unbiasedGeneration that uses `largeServer`.

File: src/Servant/QuickCheck/Internal/HasGenRequest.hs [hunk 1]
```diff
@@ -14,130 +14,138 @@ import Prelude.Compat
...
 
 instance (HasGenRequest a, HasGenRequest b) => HasGenRequest (a :<|> b) where
     genRequest _
-      = oneof [ genRequest (Proxy :: Proxy a)
-              , genRequest (Proxy :: Proxy b)
-              ]
+      = (lf + rf, frequency [l, r])
+      where
+        l@(lf, _) = genRequest (Proxy :: Proxy a)
+        r@(rf, _) = genRequest (Proxy :: Proxy b)
```
Comment (erewok) [inline src/Servant/QuickCheck/Internal/HasGenRequest.hs]: I had a test that created a list of requests and checked the paths for an API that has endpoints like this: "/1", "/2", etc. For some reason, the test would only ever return a huge list of "/1". I thought I must have been doing something wrong and I looked at it again and again.

Finally, I looked here and spotted the problem: both lines were returning `genRequest (Proxy :: Proxy a)`. After I swapped `a` for `b`, it worked as expected.

File: test/Servant/QuickCheck/InternalSpec.hs [hunk 2]
```diff
@@ -142,11 +145,29 @@ queryFlagsSpec = describe "QueryFlags" $ do
...
+unbiasedGenerationSpec = describe "Unbiased Generation of requests" $
+
+  it "frequency paired with generated endpoint should be more randomly distributed" $ do
+    let burl = BaseUrl Http "localhost" 80 ""
+    let runs = 10000 :: Double
+    someRequests <- replicateM 10000 (makeRandomRequest largeApi burl)
+    let mean = (sum $ map fromIntegral someRequests) / runs
+    let variancer x = let ix = fromIntegral x in (ix - mean) * (ix - mean)
+    let variance = (sum $ map variancer someRequests) / runs - 1
+    (variance > 20) `shouldBe` True
```
Comment (erewok) [inline test/Servant/QuickCheck/InternalSpec.hs]: So, this is likely to be a flaky test, but I wasn't sure what a good statistical test for a random distribution here would be. I first thought of variance, but then kind of guessed what a good variance metric would be. 

Does anyone have any good suggestions for testing the uniformity of the distribution of `someRequests` (which is just a giant list of numbers produced by the `HasGenRequest` instance for APIs.

**Comment by erewok** [comment]
Did travis fail the build because of missing documentation? It looks like all the test passed in the build logs.

**Comment by erewok** [comment]
I was mistaken: the travis build failed due to my potentially flaky test. I will revise it.

**Pull request merged**